codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java

```diff
@@ -327,15 +327,17 @@
     /**
      * Indicates that a {@code GOAWAY} was received from the remote endpoint and sets the last known stream.
      */
-    void goAwayReceived(int lastKnownStream, long errorCode, ByteBuf message);
+    void goAwayReceived(int lastKnownStream, long errorCode, ByteBuf message) throws Http2Exception;
 
     /**
      * Indicates whether or not a {@code GOAWAY} was sent to the remote endpoint.
      */
     boolean goAwaySent();
 
     /**
-     * Indicates that a {@code GOAWAY} was sent to the remote endpoint and sets the last known stream.
+     * Updates the local state of this {@link Http2Connection} as a result of a {@code GOAWAY} to send to the remote
+     * endpoint.
+     * @return {@code true} if the corresponding {@code GOAWAY} frame should be sent to the remote endpoint.
      */
-    void goAwaySent(int lastKnownStream, long errorCode, ByteBuf message);
+    boolean goAwaySent(int lastKnownStream, long errorCode, ByteBuf message) throws Http2Exception;
```

Comment:
note this API change. I think it makes more sense to have the connection validate these conditions, and is consistent with other connection level validation. Having the validation outside can lead to the issue fixed in this PR (the state was updated at the wrong time).